doc/tutorials/basic-validators.rst

```diff
@@ -0,0 +1,100 @@
+- The :term:`datum`, which is some script-specific data specified by the party who created the output.
+- The :term:`validation context`, which contains a representation of the spending transaction, as well as the index of the input whose validator is currently being run.
+
+These three are passed as *arguments* to the validator.
+But how are they passed?
+At least the redeemer and datum can be of different types depending on the script.
+
+The answer is that we pass them as a generic structured data type :hsobj:`Language.PlutusTx.Data.Data`.
+``Data`` is designed to make it easy to encode structured data into it, and to be itself encoded as CBOR.
+
+Consequently, the validator scripts we will write in this tutorial take three arguments of type ``Data``.
+
+However, you will typically not want to use ``Data`` directly in your program, rather you will want to use your own datatypes.
+We can easily convert to and from ``Data`` with the :hsobj:`Language.PlutusTx.IsData.Class.IsData` typeclass.
+
+You usually don't need to write your own ``IsData`` instances.
+Instead, you can use the ``makeIsData`` Template Haskell function to generate one.
+
+.. literalinclude:: BasicValidators.hs
+   :start-after: BLOCK1
+   :end-before: BLOCK2
+
+Signaling failure
+-----------------
+
+The most important thing that a validator can do is *fail*.
+This indicates that the attempt to spend the output is invalid and that transaction validation should fail.
+A validator succeeds if it does not explicitly fail.
+The actual value returned by the validator is irrelevant.
+
+How does a validator fail?
+It does so by using the :hsobj:`Language.PlutusTx.Builtins.error` builtin.
+Some other builtins may also trigger failure if they are used incorrectly (e.g. ``1/0``).
+
+Validator functions
+-------------------
+
+We write validator scripts as Haskell functions, which we compile with Plutus Tx into Plutus Core.
+The type of a validator function is ``Data -> Data -> Data -> ()``, that is, a function which takes three arguments of type ``Data``, and returns a value of type ``()`` ("unit" or "the empty tuple" -- since the return type doesn't matter we just pick something trivial).
+
+Here are two examples of simple validators that always succeed and always fail, respectively:
+
+.. literalinclude:: BasicValidators.hs
+   :start-after: BLOCK2
+   :end-before: BLOCK3
+
+If we want to write a validator that uses types other than ``Data``, we'll need to use the functions from ``IsData`` to decode them.
+Importantly, ``fromData`` can fail: in our example if the ``Data`` in the second argument is *not* a correctly encoded ``Date`` then it will return ``Nothing``, indicating that it couldn't decode it.
+However, a decoding failure indicates a mistake in the transaction that was submitted, and so we can simply fail the validation.
+
+.. important::
+   Unfortunately there's no way to provide failure diagnostics when a validator fails on chain -- it just fails.
```

Comment:
Maybe say something about how this isn't as bad as it sounds because validation is deterministic, so we can be confident that it won't fail unexpectedly on chain.

Reply:
Yep